### PR TITLE
[WIP] The nghttp2 glue code

### PR DESCRIPTION
**Pull request by jasnell**
This is a work in progress... I will update this as I go.

This starts building the process.binding('http2') API layer... which is, more or less, a shim for the nghttp2 library API. This is not yet complete.

### CODE REVIEW

File: lib/internal/http2.js [hunk 1]
```diff
@@ -0,0 +1,812 @@
...
+const kChunks = Symbol('chunks');
+const kProvider = Symbol('provider');
+const kPaused = Symbol('paused');
+const kResume = Symbol('resume');
+const kBeginResponse = Symbol('begin-response');
+const kEndStream = Symbol('end-stream');
+const kHasTrailers = Symbol('has-trailers');
+
+Object.defineProperty(exports, 'constants', {
+  configurable: false,
```
Comment (italoacasas) [inline lib/internal/http2.js:56]: `configurable: false` is not the default value ? In that case, we need to be explicit ?

Comment (jasnell) [reply lib/internal/http2.js:56]: I generally prefer to err on the side of being explicit.

File: lib/internal/http2.js [hunk 2]
```diff
@@ -0,0 +1,812 @@
...
+      // code is the error code
+      this[kStreams].delete(stream.id);
+      process.nextTick(() => this.emit('stream-close', stream.id, code));
+    };
+
+    session[constants.CALLBACK_ONDATACHUNK] = (streamID, flags, data) => {
+      // streamID is the numeric stream ID the data belongs to
+      // flags is the data flags
+      // data is the Buffer containing the data
+      const stream = this[kStreams].get(id);
```
Comment (doug-wade) [inline lib/internal/http2.js]: streamID?

File: lib/internal/http2.js [hunk 3]
```diff
@@ -0,0 +1,812 @@
...
+    if (Array.isArray(value)) {
+      this[kTrailers].set(name, value.map((i) => String(i)));
+    } else {
+      this[kTrailers].set(name, String(value));
+    }
+    return this;
+  }
+
+  addTrailers(headers) {
+    var keys = Object.keys(headers);
```
Comment (doug-wade) [inline lib/internal/http2.js]: Might

``` javascript
for (var key in headers)
  this.setTrailer(key, headers[key]);
```

be more expressive?

File: lib/internal/http2.js [hunk 4]
```diff
@@ -0,0 +1,812 @@
...
+  writeHead(statusCode, headers) {
+    this.statusCode = statusCode;
+    const keys = Object.keys(headers);
+    for (var key of keys)
+      this.setHeader(key, headers[key]);
+    return this;
+  }
+
+  _write(chunk, encoding, callback) {
+    if (typeof chunk === 'string')
```
Comment (doug-wade) [inline lib/internal/http2.js:556]: Do we need the encoding if a chunk is a Buffer?  Or, in other words, should encoding be an optional param, as in `end` directly below?

File: lib/internal/http2.js [hunk 5]
```diff
@@ -0,0 +1,812 @@
...
+    this[kSession].sendData(this[kStream]);
+  }
+
+  end(data, encoding, callback) {
+    if (typeof encoding === 'function') {
+      callback = encoding;
+      encoding = undefined;
+    }
+    if (data) {
+      this.write(data, encoding);
```
Comment (doug-wade) [inline lib/internal/http2.js]: this._write?

File: lib/internal/http2.js [hunk 6]
```diff
@@ -0,0 +1,812 @@
...
+          stream[kRequest][kFinished] = true;
+          stream[kRequest].end();
+        }
+        process.nextTick(() => {
+          this.emit('request', stream[kRequest], stream[kResponse]);
+        });
+        break;
+      case constants.NGHTTP2_HCAT_HEADERS:
+        // ignore interstitial header blocks for now, only pay attention
+        // if the header block finishes the steam, these will be the
```
Comment (doug-wade) [inline lib/internal/http2.js]: stream*

**Review by doug-wade — COMMENTED** [review]
just a coupla things I noticed while I was reading through the pr -- per https://github.com/nodejs/http2/pull/6#issuecomment-253713691, I understand that this is WIP and it might be merged before they're addressed; I'm more than happy to address any or all myself once this lands.

File: lib/internal/http2.js (hunk 4 above)
Comment (mcollina) [reply lib/internal/http2.js:556]: This is part of the stream.Writable interface.

File: lib/internal/http2.js (hunk 5 above)
Comment (mcollina) [reply lib/internal/http2.js]: no. It is part of the stream API.

File: lib/internal/http2.js [hunk 7]
```diff
@@ -0,0 +1,812 @@
...
+  }
+
+  end(data, encoding, callback) {
+    if (typeof encoding === 'function') {
+      callback = encoding;
+      encoding = undefined;
+    }
+    if (data) {
+      this.write(data, encoding);
+      super.end(callback);
```
Comment (mcollina) [inline lib/internal/http2.js]: why are we not calling super.end(chunk, callback)?

**Comment by jasnell** [comment]
@nodejs/http2 ... I'd like to get this landed even tho it's still a long ways from being done. That would allow others to help evolve this as we go.

**Comment by jasnell** [comment]
Going to get this landed.

**Comment by jasnell** [comment]
Landed (without commit metadata) in d2f03e98...67caa7a2

**Pull request closed without merging**